third_party/amd/lib/TritonAMDGPUTransforms/OptimizeBufferOpPtr.cpp

```diff
@@ -0,0 +1,349 @@
+    load->setOperand(ptrIdx, newBasePtr);
+  }
+
+  // Perform series of checks to decide if given operation could be optimized.
+  // If optimization is possible, return filled LoadData
+  static std::optional<LoadData> analyzeLoad(Operation *loadOp,
+                                             scf::ForOp targetFor) {
+    LDBG("Analyzing: " << *loadOp);
+    Value maybeOffsetsBlockArg = getOffset(loadOp);
+    auto maybeOffsetDefOp = maybeOffsetsBlockArg.getDefiningOp();
+    if (maybeOffsetDefOp && isa<arith::AddIOp>(maybeOffsetDefOp)) {
+      for (auto &use : maybeOffsetDefOp->getUses()) {
+        auto yieldOp = dyn_cast<scf::YieldOp>(use.getOwner());
+        if (!yieldOp || yieldOp->getParentOp() != targetFor) {
+          continue;
+        }
+        auto loopBody = targetFor.getBody();
+
+        int blockOpNo =
+            use.getOperandNumber() + targetFor.getNumInductionVars();
+        maybeOffsetsBlockArg = loopBody->getArgument(blockOpNo);
+        break;
+      }
+    }
+    if (!isa<BlockArgument>(maybeOffsetsBlockArg)) {
+      LDBG("Rejected: expect load offset to be a loop argument");
+      return {};
+    }
+    auto blockArg = dyn_cast<BlockArgument>(maybeOffsetsBlockArg);
+    auto loopBlock = blockArg.getOwner();
+    auto forOp = dyn_cast<scf::ForOp>(loopBlock->getParentOp());
+    if (!forOp || forOp != targetFor) {
+      LDBG("Rejected: expect load offset to be a target loop argument");
+      return {};
+    }
+    auto basePtr = getBasePtr(loadOp);
+    auto defOpBlock = basePtr.getParentBlock();
+    if (!defOpBlock->getParentOp()->isProperAncestor(targetFor)) {
+      LDBG("Rejected: expect load base Ptr to be invariant to the loop");
+      return {};
+    }
+    auto yield = dyn_cast<scf::YieldOp>(loopBlock->getTerminator());
+    int offsetOperandNo = blockArg.getArgNumber() - forOp.getNumInductionVars();
+    auto offsetYieldOperand = yield.getOperand(offsetOperandNo);
+    auto incrementOp = offsetYieldOperand.getDefiningOp();
+    if (!isa<arith::AddIOp>(incrementOp)) {
+      LDBG("Rejected: expect arith::addi used for pointer advanceent");
+      return {};
+    }
+    Value advanceStep;
+    if (incrementOp->getOperand(0) == blockArg &&
+        incrementOp->getOperand(1) != blockArg) {
+      advanceStep = incrementOp->getOperand(1);
+    }
+    if (incrementOp->getOperand(1) == blockArg &&
+        incrementOp->getOperand(0) != blockArg) {
+      advanceStep = incrementOp->getOperand(0);
+    }
+    if (!advanceStep) {
+      LDBG("Rejected: expect arith::addi to advance same block argument as "
+           "used in load");
+      return {};
+    }
+    if (!isScalarizableValue(advanceStep)) {
+      LDBG("Rejected: ptr increment step is not supported");
+      return {};
+    }
+    Value offsetInitializer = forOp.getInitArgs()[offsetOperandNo];
+    LoadData data = {loadOp, advanceStep, Value(), offsetInitializer,
+                     incrementOp};
```

Comment:
Why not call `createScalarIncrements` here to fill in `baseIncrement`? So after this analyzeLoad function, you have all info needed for this load.